### PR TITLE
Constant and derivative

### PR DESCRIPTION
**Pull request by JDBetteridge**
# Description
Improvements to some oversights with the changes I previously made to `Constant`.
- `derivative` now substitutes the symbolic constant back into the form after taking the derivative. This allows the user to perform solves with the resultant form, which was not the case before.
- Removes superfluous numbering code in derivative, which was not necessary if I had done things correctly in the first place.
- Adds `ufl2unicode` support for `Constant` via various hackery.

### CODE REVIEW

File: firedrake/constant.py [hunk 1]
```diff
@@ -197,3 +205,32 @@ def __idiv__(self, o):
...
+    var = "💩"
+    if len(o.ufl_shape) == 1:
+        var += UC.combining_right_arrow_above
+    elif len(o.ufl_shape) > 1 and self.colorama_bold:
+        var = f"{colorama.Style.BRIGHT}{var}{colorama.Style.RESET_ALL}"
+    return f"{var}{subscript_number(i)}"
+
+
+# This monkey patches ufl2unicode support for Firedrake constants
+Expression2UnicodeHandler.firedrake_constant = _unicode_format_firedrake_constant
```
Comment (nbouziani) [inline firedrake/constant.py:223]: I don't think that's the right place to add `ufl2unicode` support for constants, see my comment [here](https://github.com/firedrakeproject/firedrake/issues/3224).

Comment (JDBetteridge) [reply firedrake/constant.py:223]: I agree, this code is disgusting. I envision my code as being a temporary workaround current issues whilst someone with more UFL skills fixes this upstream.

File: tests/regression/test_constant.py [hunk 2]
```diff
@@ -231,3 +233,33 @@ class CustomConstant(Constant):
...
+    c = Constant(5)
+    f = Function(V)
+    solution = Function(V)
+
+    a = (c**2)*inner(u, v) * dx
+    L = inner(f, v) * dx
+    solve(a == L, solution)
+
+    d = derivative(a, c, IntValue(1))
+    solve(d == L, solution)
```
Comment (pbrubeck) [inline tests/regression/test_constant.py:253]: assert solution1 == (c/2)*solution2 ?

File: tests/regression/test_constant.py [hunk 3]
```diff
@@ -231,3 +233,33 @@ class CustomConstant(Constant):
...
+
+
+def test_constant_ufl2unicode():
+    mesh = UnitIntervalMesh(1)
+    a = Constant(1.0, name="a")
+    b = Constant(2.0, name="b")
+    F = a * a * b * b * dx(mesh)
+    _ = ufl2unicode(F)
+    dFda = derivative(F, u=a, du=ufl.classes.IntValue(1))
+    dFdb = derivative(F, u=b, du=ufl.classes.IntValue(1))
```
Comment (pbrubeck) [inline tests/regression/test_constant.py:263]: Could `du` be left unspecified?

File: tests/regression/test_constant.py (hunk 2 above)
Comment (JDBetteridge) [reply tests/regression/test_constant.py:253]: This is not a good idea, in general this is a hard problem to test the equality of two expressions symbolically: https://en.wikipedia.org/wiki/Computer_algebra#Equality

Comment (connorjward) [reply tests/regression/test_constant.py:253]: `np.allclose(solution1.dat.data_ro, solution2.dat.data_ro)`?

File: tests/regression/test_constant.py (hunk 3 above)
Comment (JDBetteridge) [reply tests/regression/test_constant.py:263]: Sure

File: tests/regression/test_constant.py [hunk 4]
```diff
@@ -239,3 +241,42 @@ class CustomConstant(Constant):
...
+    solution_b = Function(V)
+
+    a = (c**2)*inner(u, v) * dx
+    L = inner(f, v) * dx
+    solve(a == L, solution_a)
+
+    d = derivative(a, c, IntValue(1))
+    solve(d == L, solution_b)
+
+    assert np.allclose(solution_b.dat.data, (c.dat.data/2)*solution_a.dat.data)
```
Comment (connorjward) [inline tests/regression/test_constant.py:264]:
```suggestion
    assert np.allclose(solution_b.dat.data_ro, float(c)/2*solution_a.dat.data_ro)
```

**Review by connorjward — CHANGES_REQUESTED** [review]
I think this is basically fine, but a better mathematician than me should really take a look at this.

I think eventually we might want to consider dropping support for differentiating wrt constants. Downstream users that rely on this functionality (hopefully not many!) could do the replacement themselves.

File: tests/regression/test_constant.py (hunk 2 above)
Comment (JDBetteridge) [reply tests/regression/test_constant.py]: I did a thing

**Review by dham — APPROVED** [review]

**Pull request merged**